core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradeService.java

```diff
@@ -0,0 +1,129 @@
+package com.sequenceiq.cloudbreak.service.upgrade;
+
+import static com.sequenceiq.cloudbreak.exception.NotFoundException.notFoundException;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.Sets;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.image.ImageSettingsV4Request;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.StackViewV4Response;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.StackViewV4Responses;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.UpgradeOption;
+import com.sequenceiq.cloudbreak.cloud.model.Image;
+import com.sequenceiq.cloudbreak.common.service.TransactionService;
+import com.sequenceiq.cloudbreak.common.service.TransactionService.TransactionExecutionException;
+import com.sequenceiq.cloudbreak.common.service.TransactionService.TransactionRuntimeExecutionException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageCatalogException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageNotFoundException;
+import com.sequenceiq.cloudbreak.domain.Blueprint;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.service.ComponentConfigProviderService;
+import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.service.image.StatedImage;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.workspace.model.User;
+import com.sequenceiq.distrox.api.v1.distrox.endpoint.DistroXV1Endpoint;
+
+@Component
+public class UpgradeService {
+
+    private static final Set<Status> UPGRADEABLE_ATTACHED_DISTRO_X_STATES = Sets.immutableEnumSet(Status.STOPPED, Status.DELETE_COMPLETED);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    @Inject
+    private ComponentConfigProviderService componentConfigProviderService;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private DistroXV1Endpoint distroXV1Endpoint;
+
+    @Inject
+    private TransactionService transactionService;
+
+    public UpgradeOption getUpgradeOptionByName(Long workspaceId, String name, User user) {
```

Comment:
To be precise, it should called getUpgradeOptionByStackName and we should rename the name parameter too